File: app/src/lib/bigNumber.ts
import i18n from "./i18n";

export const BASE_UNIT_EXPONENTS = [4, 8, 12, 16, 20, 24, 28, 32, 36, 40, 44, 48] as const;
const FALLBACK_BASE_UNITS = ["万", "亿", "兆", "京", "垓", "秭", "穰", "沟", "涧", "正", "载", "极"] as const;
const FALLBACK_REPEAT_UNIT = "极";

function getBaseUnitLabels(): string[] {
    return BASE_UNIT_EXPONENTS.map((_, index) => {
        const translated = i18n.t(`number.large_units.${index}`);
        return translated && translated !== `number.large_units.${index}`
            ? translated
            : FALLBACK_BASE_UNITS[index];
    });
}

function getRepeatUnitLabel(): string {
    const translated = i18n.t("number.large_unit_repeat");
    return translated && translated !== "number.large_unit_repeat" ? translated : FALLBACK_REPEAT_UNIT;
}

export function getLargeNumberHumanUnits(): Array<[number, string]> {
    const baseLabels = getBaseUnitLabels();
    const repeatUnit = getRepeatUnitLabel();
    const units: Array<[number, string]> = [];
    for (let repeat = 0; repeat <= 6; repeat += 1) {
        const suffix = repeatUnit.repeat(repeat);
        for (const [index, exponent] of BASE_UNIT_EXPONENTS.entries()) {
            const totalExponent = exponent + repeat * 48;
            if (totalExponent < 8 || totalExponent > 300) continue;
            units.push([totalExponent, `${baseLabels[index]}${suffix}`]);
        }
    }
    units.sort((a, b) => a[0] - b[0]);
    return units;
}

export function normalizeNumericString(value: string | number | bigint): string {
    const text = String(value ?? "").trim();
    if (!text) return "0";
    if (/^[+-]?\d+$/.test(text)) {
        return BigInt(text).toString();
    }

    const match = text.match(/^([+-]?)(\d+)(?:\.(\d+))?[eE]([+-]?\d+)$/);
    if (match) {
        const sign = match[1] === "-" ? "-" : "";
        const intPart = match[2];
        const fracPart = match[3] ?? "";
        const exponent = Number.parseInt(match[4], 10);
        const digits = `${intPart}${fracPart}`.replace(/^0+/, "") || "0";
        const scale = exponent - fracPart.length;

        if (digits === "0") return "0";
        if (scale >= 0) return `${sign}${digits}${"0".repeat(scale)}`;

        const cutoff = digits.length + scale;
        if (cutoff <= 0) return "0";
        return `${sign}${digits.slice(0, cutoff)}`;
    }

    throw new Error(`invalid numeric literal: ${text}`);
}

export function compareNumericStrings(a: string | number | bigint, b: string | number | bigint): number {
    const left = BigInt(normalizeNumericString(a));
    const right = BigInt(normalizeNumericString(b));
    if (left < right) return -1;
    if (left > right) return 1;
    return 0;
}

export function addNumericStrings(a: string | number | bigint, b: string | number | bigint): string {
    return (BigInt(normalizeNumericString(a)) + BigInt(normalizeNumericString(b))).toString();
}

export function multiplyNumericStrings(a: string | number | bigint, b: string | number | bigint): string {
    return (BigInt(normalizeNumericString(a)) * BigInt(normalizeNumericString(b))).toString();
}

export function divideNumericStrings(
    a: string | number | bigint,
    b: string | number | bigint,
    options?: {
        decimals?: number;
        trimTrailingZeros?: boolean;
    },
): string {
    const decimals = Math.max(0, options?.decimals ?? 5);
    const trimTrailingZeros = options?.trimTrailingZeros ?? true;
    const dividend = BigInt(normalizeNumericString(a));
    const divisor = BigInt(normalizeNumericString(b));
    if (divisor === 0n) {
        throw new Error("Cannot divide by zero");
    }

    const negative = (dividend < 0n) !== (divisor < 0n);
    const left = dividend < 0n ? -dividend : dividend;
    const right = divisor < 0n ? -divisor : divisor;
    const scale = 10n ** BigInt(decimals);
    let quotient = (left * scale) / right;
    const remainder = (left * scale) % right;
    if (remainder * 2n >= right) {
        quotient += 1n;
    }

    let text = quotient.toString();
    if (decimals > 0) {
        if (text.length <= decimals) {
            text = text.padStart(decimals + 1, "0");
        }
        text = `${text.slice(0, -decimals)}.${text.slice(-decimals)}`;
        if (trimTrailingZeros) {
            text = text.replace(/\.?0+$/, "");
        }
    }

    if (negative && text !== "0") {
        text = `-${text}`;
    }
    return text;
}

function formatFixed(value: bigint, exponent: number, decimals: number): string {
    const scale = 10n ** BigInt(exponent);
    let quotient = (value * (10n ** BigInt(decimals))) / scale;
    const remainder = (value * (10n ** BigInt(decimals))) % scale;
    if (remainder * 2n >= scale) {
        quotient += 1n;
    }

    if (decimals <= 0) return quotient.toString();

    let text = quotient.toString();
    if (text.length <= decimals) {
        text = text.padStart(decimals + 1, "0");
    }
    return `${text.slice(0, -decimals)}.${text.slice(-decimals)}`;
}

function formatScaledFixed(value: bigint, exponent: number, decimals: number): string {
    const scale = 10n ** BigInt(exponent);
    let quotient = (value * (10n ** BigInt(decimals))) / scale;
    const remainder = (value * (10n ** BigInt(decimals))) % scale;
    if (remainder * 2n >= scale) {
        quotient += 1n;
    }

    if (decimals <= 0) return quotient.toString();

    let text = quotient.toString();
    if (text.length <= decimals) {
        text = text.padStart(decimals + 1, "0");
    }
    return `${text.slice(0, -decimals)}.${text.slice(-decimals)}`.replace(/\.?0+$/, "");
}

export function formatLargeNumber(
    value: string | number | bigint,
    options?: {
        humanDecimals?: number;
        scientificDecimals?: number;
    },
): string {
    const humanDecimals = options?.humanDecimals ?? 2;
    const scientificDecimals = options?.scientificDecimals ?? 5;
    const normalized = normalizeNumericString(value);
    const negative = normalized.startsWith("-");
    const digits = negative ? normalized.slice(1) : normalized;
    if (digits === "0") return "0";
    if (digits.length - 1 < 8) return normalized;

    const exponent = digits.length - 1;
    const sign = negative ? "-" : "";
    const number = BigInt(digits);

    if (exponent >= 304) {
        const mantissa = formatFixed(number, exponent, scientificDecimals);
        return `${sign}${mantissa}e${exponent}`;
    }

    const humanUnits = getLargeNumberHumanUnits();
    let chosen = humanUnits[0];
    for (const unit of humanUnits) {
        if (unit[0] > exponent) break;
        chosen = unit;
    }
    const [unitExponent, unitLabel] = chosen;
    return `${sign}${formatFixed(number, unitExponent, humanDecimals)}${unitLabel}`;
}

export function formatLargeScaledNumber(
    value: bigint,
    scaleDecimals: number,
    options?: {
        humanDecimals?: number;
        scientificDecimals?: number;
    },
): string {
    const scale = 10n ** BigInt(scaleDecimals);
    const negative = value < 0n;
    const abs = negative ? -value : value;
    const whole = abs / scale;
    const fraction = abs % scale;

    if (abs === 0n) return "0";

    if (whole === 0n || whole.toString().length - 1 < 8) {
        let text = whole.toString();
        if (scaleDecimals > 0 && fraction !== 0n) {
            text = `${text}.${fraction.toString().padStart(scaleDecimals, "0")}`.replace(/\.?0+$/, "");
        }
        return negative && text !== "0" ? `-${text}` : text;
    }

    const humanDecimals = options?.humanDecimals ?? 2;
    const scientificDecimals = options?.scientificDecimals ?? 5;
    const exponent = whole.toString().length - 1;
    const sign = negative ? "-" : "";

    if (exponent >= 304) {
        const mantissa = formatScaledFixed(abs, exponent + scaleDecimals, scientificDecimals);
        return `${sign}${mantissa}e${exponent}`;
    }

    const humanUnits = getLargeNumberHumanUnits();
    let chosen = humanUnits[0];
    for (const unit of humanUnits) {
        if (unit[0] > exponent) break;
        chosen = unit;
    }
    const [unitExponent, unitLabel] = chosen;
    const quotient = formatScaledFixed(abs, unitExponent + scaleDecimals, humanDecimals);
    return `${sign}${quotient}${unitLabel}`;
}

File: backend/big_number.py
from __future__ import annotations

from typing import Union


NumberLike = Union[int, str]

_BASE_UNITS: list[tuple[int, str]] = [
    (4, "万"),
    (8, "亿"),
    (12, "兆"),
    (16, "京"),
    (20, "垓"),
    (24, "秭"),
    (28, "穰"),
    (32, "沟"),
    (36, "涧"),
    (40, "正"),
    (44, "载"),
    (48, "极"),
]

_HUMAN_UNITS: list[tuple[int, str]] = []
for repeat in range(0, 7):
    suffix = "极" * repeat
    for exponent, label in _BASE_UNITS:
        total_exponent = exponent + repeat * 48
        if total_exponent < 8 or total_exponent > 300:
            continue
        _HUMAN_UNITS.append((total_exponent, label + suffix))
_HUMAN_UNITS.sort(key=lambda item: item[0])


def _strip_leading_zeros(digits: str) -> str:
    stripped = digits.lstrip("0")
    return stripped or "0"


def _normalize_scientific_numeric_string(text: str) -> str:
    sign = ""
    if text[:1] in {"+", "-"}:
        sign = "-" if text[0] == "-" else ""
        text = text[1:]

    mantissa, exponent_text = text.lower().split("e", 1)
    exponent = int(exponent_text)

    if "." in mantissa:
        integer_part, fractional_part = mantissa.split(".", 1)
    else:
        integer_part, fractional_part = mantissa, ""

    integer_part = integer_part or "0"
    digits = _strip_leading_zeros(f"{integer_part}{fractional_part}")
    if digits == "0":
        return "0"

    scale = exponent - len(fractional_part)
    if scale >= 0:
        normalized = digits + ("0" * scale)
    else:
        cutoff = len(digits) + scale
        if cutoff <= 0:
            return "0"
        normalized = digits[:cutoff]

    normalized = _strip_leading_zeros(normalized)
    if normalized == "0":
        return "0"
    return f"-{normalized}" if sign else normalized


def normalize_numeric_string(value: NumberLike) -> str:
    if isinstance(value, int):
        return str(value)

    text = str(value or "").strip()
    if not text:
        return "0"

    if "e" in text.lower():
        return _normalize_scientific_numeric_string(text)

    if "." in text:
        sign = ""
        if text[:1] in {"+", "-"}:
            sign = "-" if text[0] == "-" else ""
            text = text[1:]
        integer_part, fractional_part = text.split(".", 1)
        integer_part = _strip_leading_zeros(integer_part or "0")
        if integer_part == "0":
            return "0"
        return f"-{integer_part}" if sign else integer_part

    return str(int(text))


def to_plain_numeric_string(value: NumberLike) -> str:
    return normalize_numeric_string(value)


def _format_fixed_from_plain_numeric_string(value: str, exponent: int, decimals: int) -> str:
    number = int(value)
    scale = 10 ** exponent
    scaled = number * (10 ** decimals)
    quotient, remainder = divmod(scaled, scale)
    if remainder * 2 >= scale:
        quotient += 1

    text = str(quotient)
    if decimals <= 0:
        return text
    if len(text) <= decimals:
        text = text.zfill(decimals + 1)
    return f"{text[:-decimals]}.{text[-decimals:]}"


def format_readable_numeric_string(
        value: NumberLike,
        *,
        human_decimals: int = 2,
        scientific_decimals: int = 5,
) -> str:
    normalized = normalize_numeric_string(value)
    negative = normalized.startswith("-")
    digits = normalized[1:] if negative else normalized
    if digits == "0":
        return "0"
    if len(digits) - 1 < 8:
        return normalized

    exponent = len(digits) - 1
    sign = "-" if negative else ""

    if exponent >= 304:
        mantissa = _format_fixed_from_plain_numeric_string(digits, exponent, scientific_decimals)
        return f"{sign}{mantissa}e{exponent}"

    unit_exponent, unit_label = max(
        (item for item in _HUMAN_UNITS if item[0] <= exponent),
        key=lambda item: item[0],
    )
    coeff = _format_fixed_from_plain_numeric_string(digits, unit_exponent, human_decimals)
    return f"{sign}{coeff}{unit_label}"


def compare_numeric_strings(a: NumberLike, b: NumberLike) -> int:
    left = int(normalize_numeric_string(a))
    right = int(normalize_numeric_string(b))
    if left < right:
        return -1
    if left > right:
        return 1
    return 0


def add_numeric_strings(a: NumberLike, b: NumberLike) -> str:
    return str(int(normalize_numeric_string(a)) + int(normalize_numeric_string(b)))


def multiply_numeric_strings(a: NumberLike, b: NumberLike) -> str:
    return str(int(normalize_numeric_string(a)) * int(normalize_numeric_string(b)))


def divide_numeric_strings(
        a: NumberLike,
        b: NumberLike,
        *,
        decimals: int = 5,
        trim_trailing_zeros: bool = True,
) -> str:
    left = int(normalize_numeric_string(a))
    right = int(normalize_numeric_string(b))
    if right == 0:
        raise ZeroDivisionError("Cannot divide by zero")

    negative = (left < 0) != (right < 0)
    left = abs(left)
    right = abs(right)
    scale = 10 ** max(0, decimals)
    scaled = left * scale
    quotient, remainder = divmod(scaled, right)
    if remainder * 2 >= right:
        quotient += 1

    text = str(quotient)
    if decimals > 0:
        if len(text) <= decimals:
            text = text.zfill(decimals + 1)
        text = f"{text[:-decimals]}.{text[-decimals:]}"
        if trim_trailing_zeros:
            text = text.rstrip("0").rstrip(".")
    if negative and text != "0":
        text = f"-{text}"
    return text
